### ISSUE TITLE
TF RPI build fails to work properly with Go API

### ISSUE THREAD

**Issue by sdeoras** [tensorflow/tensorflow#24450]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
Yes

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Linux rpi3-0 4.14.79-v7+ #1159 SMP Sun Nov 4 17:50:20 GMT 2018 armv7l GNU/Linux

- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
Nope

- TensorFlow installed from (source or binary):
Source

- TensorFlow version (use command below):
At 9fd8253129

- Python version:
Python 2.7.13

- Bazel version (if compiling from source):
```
$ bazel version
INFO: Invocation ID: e77ca562-1751-40be-9b60-addbf843411d
Build label: 0.20.0- (@non-git)
Build target: bazel-out/arm-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu Dec 13 03:56:33 2018 (1544673393)
Build timestamp: 1544673393
Build timestamp as int: 1544673393
```

- GCC/Compiler version (if compiling from source):
gcc (Raspbian 4.8.5-4) 4.8.5

- CUDA/cuDNN version:
N/A

- GPU model and memory:
N/A

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

**Describe the current behavior**
I am using Go API for TensorFlow on Raspberry PI. `go test` fails with following error mesg:
```
$ go test -v github.com/tensorflow/tensorflow/tensorflow/go
# github.com/tensorflow/tensorflow/tensorflow/go [github.com/tensorflow/tensorflow/tensorflow/go.test]
./attrs.go:173:15: type [1073741824]_Ctype_longlong larger than address space
./attrs.go:173:15: type [1073741824]_Ctype_longlong too large
FAIL	github.com/tensorflow/tensorflow/tensorflow/go [build failed]

```

**Describe the expected behavior**

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by sdeoras** [comment]
@asimshankar It is great to see that TF master is cross-compiling fine for Raspberry PI (I tried at commit `92b598a32d`). Just wondering if team is planning to bring up Go API to work properly in master. I see following errors:
```
$ go test -v github.com/tensorflow/tensorflow/tensorflow/go
# github.com/tensorflow/tensorflow/tensorflow/go
/tmp/go-build505912479/b048/_x003.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_ContextListDevices':
/tmp/go-build/cgo-gcc-prolog:44: undefined reference to `TFE_ContextListDevices'
/tmp/go-build505912479/b048/_x003.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_ContextOptionsSetAsync':
/tmp/go-build/cgo-gcc-prolog:61: undefined reference to `TFE_ContextOptionsSetAsync'
/tmp/go-build505912479/b048/_x003.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_ContextOptionsSetConfig':
/tmp/go-build/cgo-gcc-prolog:76: undefined reference to `TFE_ContextOptionsSetConfig'
/tmp/go-build505912479/b048/_x003.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_DeleteContext':
/tmp/go-build/cgo-gcc-prolog:88: undefined reference to `TFE_DeleteContext'
/tmp/go-build505912479/b048/_x003.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_DeleteContextOptions':
/tmp/go-build/cgo-gcc-prolog:100: undefined reference to `TFE_DeleteContextOptions'
/tmp/go-build505912479/b048/_x003.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_NewContext':
/tmp/go-build/cgo-gcc-prolog:116: undefined reference to `TFE_NewContext'
/tmp/go-build505912479/b048/_x003.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_NewContextOptions':
/tmp/go-build/cgo-gcc-prolog:133: undefined reference to `TFE_NewContextOptions'
/tmp/go-build505912479/b048/_x004.o: In function `_cgo_70e30ee47b92_Cfunc_TF_AddGradientsWithPrefix':
/tmp/go-build/cgo-gcc-prolog:61: undefined reference to `TF_AddGradientsWithPrefix'
/tmp/go-build505912479/b048/_x011.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_DeleteTensorHandle':
/tmp/go-build/cgo-gcc-prolog:40: undefined reference to `TFE_DeleteTensorHandle'
/tmp/go-build505912479/b048/_x011.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_NewTensorHandle':
/tmp/go-build/cgo-gcc-prolog:56: undefined reference to `TFE_NewTensorHandle'
/tmp/go-build505912479/b048/_x011.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_TensorHandleBackingDeviceName':
/tmp/go-build/cgo-gcc-prolog:75: undefined reference to `TFE_TensorHandleBackingDeviceName'
/tmp/go-build505912479/b048/_x011.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_TensorHandleCopyToDevice':
/tmp/go-build/cgo-gcc-prolog:96: undefined reference to `TFE_TensorHandleCopyToDevice'
/tmp/go-build505912479/b048/_x011.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_TensorHandleDataType':
/tmp/go-build/cgo-gcc-prolog:114: undefined reference to `TFE_TensorHandleDataType'
/tmp/go-build505912479/b048/_x011.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_TensorHandleDeviceName':
/tmp/go-build/cgo-gcc-prolog:133: undefined reference to `TFE_TensorHandleDeviceName'
/tmp/go-build505912479/b048/_x011.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_TensorHandleDim':
/tmp/go-build/cgo-gcc-prolog:153: undefined reference to `TFE_TensorHandleDim'
/tmp/go-build505912479/b048/_x011.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_TensorHandleNumDims':
/tmp/go-build/cgo-gcc-prolog:172: undefined reference to `TFE_TensorHandleNumDims'
/tmp/go-build505912479/b048/_x011.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_TensorHandleResolve':
/tmp/go-build/cgo-gcc-prolog:191: undefined reference to `TFE_TensorHandleResolve'
collect2: error: ld returned 1 exit status
FAIL	github.com/tensorflow/tensorflow/tensorflow/go [build failed]
```

**Comment by asimshankar** [comment]
@sdeoras : those messages seems symptomatic of having an older version of the C library. Could it be that you are linking against an old version?

**Comment by sdeoras** [comment]
@asimshankar : thank you. yes, it was an older library i was linking against. Previous errors are gone but one last issue still remains:
```
$ go test -v github.com/tensorflow/tensorflow/tensorflow/go
# github.com/tensorflow/tensorflow/tensorflow/go [github.com/tensorflow/tensorflow/tensorflow/go.test]
./attrs.go:173:15: type [1073741824]_Ctype_longlong larger than address space
./attrs.go:173:15: type [1073741824]_Ctype_longlong too large
FAIL	github.com/tensorflow/tensorflow/tensorflow/go [build failed]
```
I am able to get past this error simply by reducing the slice length in `attrs.go:173` to:
`slice := (*[1 << 27]C.int64_t)(unsafe.Pointer(dim))[:numDim:numDim]`

**Comment by deslum** [comment]
```
root@pc:/home/deslum/src/test# go get github.com/tensorflow/tensorflow/tensorflow/go
# github.com/tensorflow/tensorflow/tensorflow/go
/tmp/go-build375822434/b001/_x011.o: In function `_cgo_70e30ee47b92_Cfunc_TFE_TensorHandleBackingDeviceName':
/tmp/go-build/cgo-gcc-prolog:75: undefined reference to `TFE_TensorHandleBackingDeviceName'
collect2: error: ld returned 1 exit status
```

**Comment by asimshankar** [comment]
@deslum : Sorry about that, the issue you encountered is different from what @sdeoras and is more like https://github.com/tensorflow/tensorflow/issues/23257#issuecomment-433751410 (Which I agree is a problem regardless). For now, you could work around this by doing what is suggested in the comment there. I'll send out a fix so that the `master` branch compiles again. Longer term probably want to figure something out with go modules or something.

@sdeoras : Sorry, haven't been able to dig into this - from a cursory look seems like something to do with `long long*` pointers on a 32-bit architecture?

**Comment by asimshankar** [comment]
@sdeoras: Oh, right, sorry missed that last sentence of your previous message. Would you like to contribute a PR for this?

**Comment by sdeoras** [comment]
@asimshankar: sure, i'll send out a pr for this bug.

**Comment by asimshankar** [comment]
For the record, fixed by #24840
Thanks @sdeoras
